ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java

```diff
@@ -1747,15 +1747,15 @@ public void testMultiInsertOnDynamicallyPartitionedMmTable() throws Exception {
     final String completedTxnComponentsContents =
         TxnDbUtil.queryToString(conf, "select * from \"COMPLETED_TXN_COMPONENTS\"");
     Assert.assertEquals(completedTxnComponentsContents,
-        2, TxnDbUtil.countQueryAgent(conf, "select count(*) from \"COMPLETED_TXN_COMPONENTS\""));
+        4, TxnDbUtil.countQueryAgent(conf, "select count(*) from \"COMPLETED_TXN_COMPONENTS\""));
```

Comment:
Is this change expected? What are these records?

Reply:
Those records are duplicates. It is a "side-effect" of fixing the FileSinkOperator-MoveTask assignment.
For ACID tables for an insert like in the test, 4 records were created even before the direct insert got introduced. Because then the FSO-MoveTask assignment was based on the staging directories. And for insert like this there were 2 FSOs and 2 MoveTasks. Each MoveTasks called the metastore method which creates an entry in the TXN_COMPONENTS table for each partition. So there were 4 records at the end of the insert. But for MM tables (and later for direct insert) there is no staging directory and all MoveTasks and all FSOs will contain the table directory. So for every FSO it will find the same MoveTask (which is the first in the list) and only this one will be executed. This is not correct, but didn't cause any issue, so it was undetected until the direct delete and update came in. To make them work properly, had to fix the FSO-MoveTask assignment, but then for MM tables and with direct insert it will have duplicate records just like for ACID tables without direct insert. The Java doc of the TxnHandler.addDynamicPartitions method says that duplicates won't cause any trouble, but if you know issues with that, please share it with me.